youtube_dl/extractor/panopto.py

```diff
@@ -0,0 +1,296 @@
+
+    def _real_extract(self, url):
+        """Extracts the video and stream information for the given Panopto hosted URL."""
+        video_id = self._match_id(url)
+        org = self._match_organization(url)
+
+        delivery_info = self._download_json(
+            'https://{}.hosted.panopto.com/Panopto/Pages/Viewer/DeliveryInfo.aspx'.format(org),
+            video_id,
+            query={
+                'deliveryId': video_id,
+                'invocationId': '',
+                'isLiveNotes': 'false',
+                'refreshAuthCookie': 'true',
+                'isActiveBroadcast': 'false',
+                'isEditing': 'false',
+                'isKollectiveAgentInstalled': 'false',
+                'isEmbed': 'false',
+                'responseType': 'json',
+            }
+        )
+
+        if 'ErrorCode' in delivery_info:
+            self._downloader.report_warning("If the video you are trying to download requires you to sign-in, you will "
+                                            "need to provide a cookies file that allows the downloader to authenticate "
+                                            "with Panopto. If the error below is about unauthorized access, this is "
+                                            "most likely the issue.")
+            raise ExtractorError(
+                'API error: ({}) {}'.format(delivery_info.get('ErrorCode', '?'), delivery_info.get('ErrorMessage', '?'))
+            )
+
+        streams = []
+        for this_stream in delivery_info['Delivery']['Streams']:
+            new_stream = {
+                'id': this_stream['PublicID'],
+                'title': this_stream['Tag'],
+                'formats': [],
+            }
+            if 'StreamHttpUrl' in this_stream:
+                new_stream['formats'].append({
+                    'url': this_stream['StreamHttpUrl'],
+                })
+            if 'StreamUrl' in this_stream:
+                m3u8_formats = self._extract_m3u8_formats(this_stream['StreamUrl'], video_id, 'mp4')
+                self._sort_formats(m3u8_formats)
+                new_stream['formats'].extend(m3u8_formats)
+            if len(new_stream['formats']):
+                streams.append(new_stream)
+
+        if not streams:
+            raise ExtractorError('No streams found.')
+
+        result = {
+            'id': video_id,
+            'title': delivery_info['Delivery']['SessionName'],
+            'thumbnail': 'https://{}.hosted.panopto.com/Panopto/Services/FrameGrabber.svc/FrameRedirect?objectId={}&mode=Delivery&random={}'.format(
+                         org, video_id, random()),
+        }
+
+        if len(streams) == 1:
+            result['formats'] = streams[0]['formats']
+        else:
+            result['_type'] = 'multi_video'
+            result['entries'] = streams
+
+        # We already know Delivery exists since we need it for stream extraction
+        contributors = delivery_info['Delivery'].get('Contributors')
```

Comment:
I'm not sure how this is at odds with the coding conventions. Are you referring to the use of `['Delivery']`? The Delivery key _must_ exist assuming that code is ever executed because the key is needed to retrieve non-optional fields earlier. It would fail at line 120:

```python
for this_stream in delivery_info['Delivery']['Streams']:
```

Because we need `['Delivery']` (and more specifically `['Delivery']['Streams']`) to even find the non-optional information.